File: api/app/schemas/game.py
"""
Pydantic schemas for game API end points.
"""
from pydantic import BaseModel, Field
from typing import Dict, List, Optional
from datetime import date
from enum import Enum


# Define enums locally to avoid import issues
class RecommendationType(str, Enum):
    STRONG_BUY = "STRONG_BUY"
    BUY = "BUY"
    HOLD = "HOLD"
    SELL = "SELL"
    STRONG_SELL = "STRONG_SELL"


class SignalType(str, Enum):
    BULLISH = "BULLISH"
    NEUTRAL = "NEUTRAL"
    BEARISH = "BEARISH"


class RiskLevel(str, Enum):
    LOW = "LOW_RISK"
    MEDIUM = "MEDIUM_RISK"
    HIGH = "HIGH_RISK"


class GamePrice(BaseModel):
    """OHLCV prices for a single ticker on a single day."""
    open: float = Field(..., description="Opening price")
    high: float = Field(..., description="Highest price")
    low: float = Field(..., description="Lowest price")
    close: float = Field(..., description="Closing price")
    volume: int = Field(..., description="Trading volume")


class GameRecommendation(BaseModel):
    """Simplified recommendation for game UI."""
    ticker: str = Field(..., description="Stock ticker symbol")
    recommendation: RecommendationType = Field(..., description="AI recommendation")
    confidence: float = Field(..., ge=0, le=1, description="Confidence score (0-1)")
    technical_signal: SignalType = Field(..., description="Technical analysis signal")
    sentiment_signal: SignalType = Field(..., description="Sentiment analysis signal")
    risk_level: RiskLevel = Field(..., description="Risk assessment")
    rationale_summary: str = Field(..., description="Short explanation for students")
    rationale_risk_view: List[str] = Field(default_factory=list, description="Risk-related reasoning (bullet points)")
    rationale_technical_view: List[str] = Field(default_factory=list, description="Technical analysis reasoning (bullet points)")
    rationale_sentiment_view: List[str] = Field(default_factory=list, description="Sentiment analysis reasoning (bullet points)")
    as_of_date: str = Field(..., description="ISO date the AI used for analysis")


class NewsArticle(BaseModel):
    """News article for a ticker on a specific day."""
    ticker: str = Field(..., description="Stock ticker symbol")
    headline: str = Field(..., description="News headline")
    content: Optional[str] = Field(None, description="Article content/summary")
    source: str = Field(..., description="News source")
    published_at: str = Field(..., description="Publication timestamp")
    sentiment_score: Optional[float] = Field(None, ge=-1, le=1, description="Sentiment score (-1 to +1)")
    url: Optional[str] = Field(None, description="Article URL")


class GameDayResponse(BaseModel):
    """Complete data for a single game day."""
    day: int = Field(..., description="Day number (0-indexed)")
    date: str = Field(..., description="ISO date string")
    is_trading_day: bool = Field(..., description="Whether markets are open (weekdays)")
    recommendations: List[GameRecommendation] = Field(..., description="AI recommendations for all tickers")
    prices: Dict[str, GamePrice] = Field(..., description="OHLC prices for all tickers")
    news: List[NewsArticle] = Field(default_factory=list, description="News articles for this day")


class GameDataResponse(BaseModel):
    """Complete game data for N days."""
    days: List[GameDayResponse] = Field(..., description="All game days")
    tickers: List[str] = Field(..., description="List of ticker symbols")
    start_date: str = Field(..., description="Game start date (ISO)")
    end_date: str = Field(..., description="Game end date (ISO)")
    total_days: int = Field(..., description="Total number of days")

File: web/types/game.ts
/**
 * TypeScript types for the educational stock game
 */

import type { RecommendationType, SignalType, RiskLevel } from './api';

/**
 * Technical indicators for a ticker on a specific day
 */
export interface TechnicalIndicators {
  sma_20?: number;
  sma_50?: number;
  sma_200?: number;
  ema_12?: number;
  ema_26?: number;
  rsi_14?: number;
  macd?: number;
  macd_signal?: number;
  macd_histogram?: number;
  bollinger_upper?: number;
  bollinger_middle?: number;
  bollinger_lower?: number;
  atr_14?: number;
  obv?: number;
  volatility_30d?: number;
}

/**
 * A single day's market data and AI recommendations
 */
export interface GameDay {
  day: number; // 0-indexed (0 = Day 1)
  date: string; // ISO date string
  is_trading_day: boolean; // Whether markets are open (weekdays)
  recommendations: DayRecommendation[];
  prices: Record<string, DayPrice>; // ticker -> prices
  news?: NewsArticle[]; // News articles for this day
  technical_indicators?: Record<string, TechnicalIndicators>; // ticker -> technical indicators
}

/**
 * AI recommendation for a specific ticker on a specific day
 */
export interface DayRecommendation {
  ticker: string;
  recommendation: RecommendationType;
  confidence: number;
  technical_signal: SignalType;
  sentiment_signal: SignalType;
  risk_level: RiskLevel;
  rationale_summary: string; // Short explanation for students
  rationale_technical_view: string[];
  rationale_sentiment_view: string[];
  rationale_risk_view: string[];
  as_of_date: string; // ISO date the AI used for analysis
}

/**
 * OHLCV prices for a ticker on a specific day
 */
export interface DayPrice {
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

/**
 * News article for a ticker on a specific day
 */
export interface NewsArticle {
  ticker: string;
  headline: string;
  content?: string | null;
  source: string;
  published_at: string;
  sentiment_score?: number | null; // -1 to +1
  url?: string | null;
}

/**
 * Complete game data loaded at start
 */
export interface GameData {
  days: GameDay[];
  tickers: string[];
  start_date: string; // API returns snake_case
  end_date: string; // API returns snake_case
  total_days: number; // API returns snake_case
}

/**
 * A trade executed by the player or AI
 */
export interface Trade {
  id: string;
  day: number;
  date: string;
  ticker: string;
  type: 'BUY' | 'SELL';
  shares: number;
  price: number; // Execution price
  total: number; // shares * price
  portfolioValue: number; // Total portfolio value after trade
}

/**
 * Holdings for a specific ticker
 */
export interface Holding {
  ticker: string;
  shares: number;
  avgCost: number; // Average cost per share
  totalCost: number; // Total invested
  currentValue: number; // Current market value
  unrealizedPnL: number; // Profit/loss
  unrealizedPnLPercent: number; // Profit/loss percentage
}

/**
 * Portfolio snapshot at end of day
 */
export interface PortfolioSnapshot {
  day: number;
  date: string;
  cash: number;
  holdingsValue: number;
  totalValue: number;
  returnPercent: number;
  returnDollars: number;
}

/**
 * Score breakdown
 */
export interface ScoreBreakdown {
  portfolioReturn: number; // points from return vs. benchmark
  riskDiscipline: number; // points from following AI signals
  beatAI: number; // Bonus points for beating AI
  drawdownPenalty: number; // Penalty for large drawdowns
  totalScore: number;
  grade: string; // A+ to F
}

/**
 * Game configuration
 */
export interface GameConfig {
  initialCash: number; // Default: $100,000
  numDays: number; // Default: 30
  tickers: string[]; // Default: ['AAPL', 'MSFT', 'GOOGL', 'AMZN']
  difficulty: 'easy' | 'medium' | 'hard'; // Affects scoring thresholds
  startDate?: string; // Optional start date for simulations
}

/**
 * Player's game state
 */
export interface PlayerState {
  // Identity
  playerId: string;
  playerName: string;

  // Game progress
  currentDay: number; // 0-indexed
  gameStartedAt: string; // ISO timestamp
  gameEndedAt?: string; // ISO timestamp (when finished)

  // Portfolio
  cash: number;
  holdings: Record<string, Holding>; // ticker -> holding
  trades: Trade[];

  // History
  portfolioHistory: PortfolioSnapshot[];

  // Score
  score: number;
  scoreBreakdown: ScoreBreakdown;
  grade: string;
}

/**
 * AI opponent state (runs in parallel)
 */
export interface AIState {
  cash: number;
  holdings: Record<string, {
    ticker: string;
    shares: number;
    avgCost: number;
    totalCost: number;
    currentValue: number;
    unrealizedPnL: number;
    unrealizedPnLPercent: number;
  }>;
  trades: Trade[];
  portfolioHistory: PortfolioSnapshot[];
}

/**
 * Complete game state
 */
export interface GameState {
  // Game data
  config: GameConfig;
  gameData: GameData | null;
  isLoading: boolean;

  // Player state
  player: PlayerState;

  // AI opponent state
  ai: AIState;

  // Game status
  status: 'not_started' | 'playing' | 'finished';

  // Multiplayer (Phase 4)
  roomCode?: string;
  isMultiplayer: boolean;
  role?: 'teacher' | 'student' | null; // User's role in multiplayer
}

/**
 * Trade validation result
 */
export interface TradeValidation {
  allowed: boolean;
  reason?: string;
}

/**
 * Multiplayer room (Phase 4)
 */
export interface GameRoom {
  roomCode: string;
  createdBy: string; // Teacher name
  createdAt: string;
  config: GameConfig;
  players: PlayerState[];
  status: 'waiting' | 'in_progress' | 'finished';
}

/**
 * Leaderboard entry
 */
export interface LeaderboardEntry {
  rank: number;
  playerId: string;
  playerName: string;
  score: number;
  grade: string;
  portfolioValue: number;
  returnPercent: number;
  currentDay: number;
}
